### PR TITLE
feat: add message header

### PR DESCRIPTION
**Pull request by bhavya2611**
Note - This is a draft PR and WIP. Will be merging this post current open PR's are merged 

## What does this PR do?

Adds message header component and also infinite scroll for messages

## Type of change

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Enhancement (non-breaking small changes to existing functionality)

## How should this be tested?

- [ ] Renders messages correctly and loads 20 at a time
- [ ] Message Header component renders correctly

Screenshots

image

### CODE REVIEW

File: src/components/Shared/MessagesList.tsx
Comment (elisealix22) [inline src/components/Shared/MessagesList.tsx:92]: Is it possible to reuse the `useInView` hook that powers pagination in places like the Explore feed instead of introducing a new `InfiniteScroll` dep?

https://github.com/lensterxyz/lenster/blob/c76564fb8672bf3a99f8bd9295e76cb8354dcad8/src/components/Explore/Feed.tsx#L42-L54

File: src/components/utils/hooks/useGetMessages.tsx
Comment (elisealix22) [inline src/components/utils/hooks/useGetMessages.tsx:18]: Unsure if this is just for testing, but I think we can likely bump this page size up a bit -- maybe something like last 20 messages?

Comment (neekolas) [reply src/components/utils/hooks/useGetMessages.tsx:18]: 20 seems like a reasonable number to start. We can play with it using some large conversations and see how it feels.

File: src/components/Shared/MessageHeader.tsx
Comment (neekolas) [inline src/components/Shared/MessageHeader.tsx:10]: Shouldn't this be the other person's profile, not your own?

File: src/components/Shared/MessageHeader.tsx
Comment (neekolas) [inline src/components/Shared/MessageHeader.tsx:9]: Seems like this needs to get the initial following state from the API. The user may be already followed prior to loading the component.

File: src/components/Messages/Message.tsx
Comment (neekolas) [inline src/components/Messages/Message.tsx:33]: This should be memoized, since it gets passed to a component.

File: src/components/Shared/MessageHeader.tsx
Comment (neekolas) [inline src/components/Shared/MessageHeader.tsx:8]: I wonder if this would be better as a stateless component that takes data in as props. Similar to the `MessagesList`. Would be nice to handle all the state management in the parent.

File: src/components/utils/hooks/useGetMessages.tsx
Comment (elisealix22) [reply src/components/utils/hooks/useGetMessages.tsx:18]: Also @bhavya2611 @neekolas , is it possible to use `await conversation.messagesPaginated({ pageSize: 20 })` here so we can page thru messages without having to query by end time?

Comment (neekolas) [reply src/components/utils/hooks/useGetMessages.tsx:18]: I was trying to figure out a nice way to do that. It's a little messy trying to pause an async generator. You have to await some Promise in the loop, and have that promise get resolved externally and replaced with a new promise before the next loop. Otherwise it will just page until the end. It's possible, but not pretty.

File: src/components/Shared/MessageHeader.tsx
Comment (bhavya2611) [reply src/components/Shared/MessageHeader.tsx:10]: Yes, was waiting for Elise's PR to get merged, this is all sorted now

File: src/components/Shared/MessageHeader.tsx
Comment (bhavya2611) [reply src/components/Shared/MessageHeader.tsx:9]: Yes, was waiting for Elise's PR to get merged, this is all sorted now

File: src/components/Shared/MessageHeader.tsx
Comment (bhavya2611) [reply src/components/Shared/MessageHeader.tsx]: Resolved

File: src/components/Messages/Message.tsx
Comment (bhavya2611) [reply src/components/Messages/Message.tsx]: Do you want me to use, useCallback here?

because memo would only make sense if it had an input, and it does not have one

File: src/components/Shared/MessagesList.tsx
Comment (bigint) [reply src/components/Shared/MessagesList.tsx]: Yea we don't want to add external dependency for the same purpose 🙏🏼

File: src/components/Messages/Message.tsx
Comment (neekolas) [reply src/components/Messages/Message.tsx]: Yes, exactly

File: src/components/Shared/MessagesList.tsx
Comment (bhavya2611) [reply src/components/Shared/MessagesList.tsx]: This seems to not work well, I tried. Will further look into it

File: src/components/Messages/Message.tsx
Comment (bhavya2611) [reply src/components/Messages/Message.tsx]: Okay this is done

File: src/components/Shared/MessagesList.tsx
Comment (bhavya2611) [reply src/components/Shared/MessagesList.tsx]: Resolved this

Comment (bigint) [reply src/components/Shared/MessagesList.tsx]: Amazing @bhavya2611

File: src/components/utils/hooks/useGetMessages.tsx
Comment (elisealix22) [inline src/components/utils/hooks/useGetMessages.tsx:30]: Might be nice to pull the `20` out to its own constant so it stays in sync if someone changes the page size on line 18

File: src/components/Shared/MessageHeader.tsx
Comment (elisealix22) [inline src/components/Shared/MessageHeader.tsx:25]: Can we wrap this in a ```<Link href={`/u/${profile?.handle}`}>``` so we can click through to the person from the header?

**Review by elisealix22 — COMMENTED** [review]
Nice! I noticed a few things that might be worth looking into in testing:
1. The list scrolling seems to become quite choppy when fetching a new page. Is it possible there's some work being done synchronously?
2. When I hit the bottom of a page, it stays like that for all convos so I can't paginate even if I change to another convo. To repro, you'll need two convos that paginate.
3. One of my convos is stuck in a Loading state at the top (first message) somehow. I'm thinking maybe it's last page size was exactly 20? IdkIdk but it's reproducible.

<details>
<summary>Premature end of convo</summary>

Screen Shot 2022-10-13 at 4 51 29 PM

</details>

<details>
<summary>Stuck loading</summary>
Screen Shot 2022-10-13 at 4 44 59 PM
</details>

File: src/components/Shared/MessageHeader.tsx
Comment (bhavya2611) [reply src/components/Shared/MessageHeader.tsx:20]: My bad, removed it

File: src/components/Messages/Message.tsx
Comment (bhavya2611) [reply src/components/Messages/Message.tsx:37]: Done

File: src/components/utils/hooks/useGetMessages.tsx
Comment (bhavya2611) [reply src/components/utils/hooks/useGetMessages.tsx]: Done

File: src/components/Shared/MessageHeader.tsx
Comment (bhavya2611) [reply src/components/Shared/MessageHeader.tsx]: Done

**Comment by bigint** [comment]
image

scrollbar is overlapping the message tile 😕

**Comment by bigint** [comment]
Screenshot 2022-10-14 at 6 05 25 PM

It would be great to show the complete user profile here, something like this

Screenshot 2022-10-14 at 6 05 48 PM

**Review by elisealix22 — APPROVED** [review]
This is awesome Bhavya. It's still freezing up for me a bit when paging (attached a video). But I'm happy to dig into that in a follow up if you prefer!

<details>
<summary>Paging example</summary>

https://user-images.githubusercontent.com/556051/195857681-d30876b0-294f-47b6-96ee-c04b0c40ffbc.mov

</details>

**Comment by elisealix22** [comment]
Ooooh one more thing came up in the demo today: if I tapped 'Follow' the button disappeared. I think we may want to show unfollow there as well to be able to undo it.

**Comment by bigint** [comment]
@bhavya2611 @elisealix22 I'm seeing two scrollbars when using with an external mouse 😕 

image

**Comment by bhavya2611** [comment]
Resolved scroll overlap issue and message header show profile issue, the other one about two scroll bars will be fixed when we solve for overall height of the page

**Comment by bhavya2611** [comment]
I have also added unfollow button, it seems to have a longer flow, also @bigint do I need to refetch the data after follow/unfollow for the state to reflect the latest?

File: src/components/Shared/MessagesList.tsx
Comment (bigint) [reply src/components/Shared/MessagesList.tsx]: @bhavya2611 We need to move back to `react-infinite-scroller` we came to know `usInview` is more buggy and complex to maintain

We suggest you revert back the changes to use `react-infinite-scroller`

Sorry for the inconvenience happened 🙏🏼 

Make sure to rebase the branch 😊

**Comment by bigint** [comment]
@bhavya2611 You can't refetch the data cuz it will take some time to index in the blockchain

Better to optimistically show unfollow button after clicking the follow as we have on the profile page. It's ok to have a longer flow 🙇🏼

**Comment by bhavya2611** [comment]
@bigint I saw you are using react-infinite-scroller, this is very buggy I feel, it is absolutely not waiting for the scroll and just off the go starts calling the API, I feel this is a better library -> react-infinite-scroll-component and we should use this

**Comment by bigint** [comment]
@bhavya2611 I think u need to set `useWindow` to false 🤔 

It is working fine in other places 👀

**Comment by bhavya2611** [comment]
Okay, I spent a lot of time trying to make this react-infinite-scroll thing work correctly also tried with virtualized list stuff, it is all very buggy. I wanted to ensure we get done with all functionality first and then get back to this, so I removed it for now. I will work on the new conversation flow and then on this. Let's merge this PR if you guys feel the approach is okay

@bigint @elisealix22 @neekolas

**Comment by bhavya2611** [comment]
I think the main problem is this list we want in reverse direction and that is not being picked up correctly

**Review by bigint — APPROVED** [review]
**LGTM** :shipit: 

Rad LGTM

**Pull request merged**
